examples/threading/threading.c

```diff
@@ -2,15 +2,51 @@
 #include <unistd.h>
 #include <stdlib.h>
 #include <stdio.h>
+#include <errno.h>
+#include <time.h>
 
 // Optional: use these functions to add debug or error prints to your application
-#define DEBUG_LOG(msg,...)
-//#define DEBUG_LOG(msg,...) printf("threading: " msg "\n" , ##__VA_ARGS__)
+//#define DEBUG_LOG(msg,...)
+#define DEBUG_LOG(msg,...) printf("threading: " msg "\n" , ##__VA_ARGS__)
 #define ERROR_LOG(msg,...) printf("threading ERROR: " msg "\n" , ##__VA_ARGS__)
 
 void* threadfunc(void* thread_param)
 {
+    struct thread_data* thread_func_args = (struct thread_data *) thread_param;
 
+    int ret = usleep(thread_func_args->wait_to_obtain_ms * 1000);
+
+    if (ret == -1) {
+        ERROR_LOG("Error in usleep with code: %d", errno);
+        thread_func_args->thread_complete_success = false;
+        thread_func_args->error_status = 1;
+        return thread_param;
+    } 
+
+    int rc = pthread_mutex_lock(thread_func_args->mutex);
+    if (rc != 0) {
+        ERROR_LOG("Error in mutex lock with code: %d", errno);
+        thread_func_args->thread_complete_success = false;
+        thread_func_args->error_status = 1;
```

Comment:
@GautamaGandhi rather than duplicating the cleanup code here and in every failure branch, I suggest handling errors and returning in a shared code path

Reply:
@dwalkes noted. I guess another approach would be to use a local variable for error status and use the cleanup logic after a check on this variable. But, in the event that the first function call usleep() fails, we would want to set the complete_status to false and return immediately right? Would a function which only performs cleanup be better for this? Something like cleanup(void* thread_param)